src/resolve/resolved-dns-transaction.c

```diff
@@ -1117,13 +1119,128 @@ void dns_transaction_process_reply(DnsTransaction *t, DnsPacket *p, bool encrypt
+                        return;
+                }
+                if (r < 0) {
+                        /* On LLMNR, if we cannot connect to the host,
+                         * we immediately give up */
+                        if (t->scope->protocol != DNS_PROTOCOL_DNS)
+                                goto fail;
+
+                        /* On DNS, couldn't send? Try immediately again, with a new server */
+                        if (dns_transaction_limited_retry(t))
+                                return;
+
+                        /* No new server to try, give up */
+                        dns_transaction_complete(t, DNS_TRANSACTION_ATTEMPTS_MAX_REACHED);
+                }
+
+                return;
+        }
+
+        /* After the superficial checks, actually parse the message. */
+        r = dns_packet_extract(p);
+        if (r < 0) {
+                if (t->server) {
+                        dns_server_packet_invalid(t->server, t->current_feature_level);
+
+                        r = dns_transaction_maybe_restart(t);
+                        if (r < 0)
+                                goto fail;
+                        if (r > 0) /* Transaction got restarted... */
+                                return;
+                }
+
+                dns_transaction_complete(t, DNS_TRANSACTION_INVALID_REPLY);
+                return;
+        }
+
         switch (t->scope->protocol) {
 
         case DNS_PROTOCOL_DNS:
                 assert(t->server);
 
+                r = dns_packet_ede_rcode(p, &ede_msg);
+                if (r < 0 && r != -EINVAL)
+                        goto fail;
+                ede_rcode = r;
+
                 if (!t->bypass &&
                     IN_SET(DNS_PACKET_RCODE(p), DNS_RCODE_FORMERR, DNS_RCODE_SERVFAIL, DNS_RCODE_NOTIMP)) {
+                        /* If the server has replied with detailed error data, using a degraded feature set
+                         * will likely not help anyone. Examine the detailed error to determine the best
+                         * course of action. */
+                        if (ede_rcode >= 0 && DNS_PACKET_RCODE(p) == DNS_RCODE_SERVFAIL) {
+                                /* These codes are related to DNSSEC configuration errors. If accurate,
+                                 * this is the domain operator's problem, and retrying won't help. */
+                                if (dns_ede_rcode_is_dnssec(ede_rcode)) {
+                                        log_debug("Server returned error: %s (%s%s%s). Lookup failed.",
+                                                        FORMAT_DNS_RCODE(DNS_PACKET_RCODE(p)),
+                                                        FORMAT_DNS_EDE_RCODE(ede_rcode),
+                                                        isempty(ede_msg) ? "" : ": ", ede_msg);
+                                        dns_transaction_complete(t, DNS_TRANSACTION_DNSSEC_FAILED);
```

Comment:
This isn't really accurate. I think I will need a new transaction result type here, or should I just stuff something sensible in answer_dnssec_result so the log messages don't break?

Reply:
yeha, I think it would be nice to add another enum value to DnssecResult and set that in answer_dnssec_result

Reply:
I ended up adding a new transaction result to mark the upstream error instead. I don't think it was quite flexible enough to just re-use dnssec result. So now it can report the upstream message to the user:

```
[root@arch ~]# resolvectl query dnssec-failed.org
dnssec-failed.org: resolve call failed: DNSSEC validation failed upstream: DNSKEY Missing: no SEP matching the DS found for dnssec-failed.org.
```

It's a little verbose in this format but that's probably fine I think. PTAL.